src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -0,0 +1,111 @@
+-- Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Daml.Finance.Instrument.Swap.Fpml.Util where
+
+import ContingentClaims.Claim (Inequality(..), give, one, scale, when)
+import ContingentClaims.Observation (Observation(..))
+import DA.Date
+import DA.Optional (fromOptional, fromSome, isNone, isSome)
+import Daml.Finance.Instrument.Generic.Util
+import Daml.Finance.Interface.Claims.Types (Deliverable, Observable, TaggedClaim(..))
+import Daml.Finance.Interface.Types.Date.Calendar
+import Daml.Finance.Interface.Types.Date.DayCount (DayCountConventionEnum)
+import Daml.Finance.Interface.Types.Date.RollConvention
+import Daml.Finance.Interface.Types.Date.Schedule (PeriodicSchedule(..), SchedulePeriod, StubPeriodTypeEnum(..))
+import Daml.Finance.Interface.Types.Fpml
+import Daml.Finance.Util.Date.Calendar (addBusinessDays, adjustDate, merge)
+import Daml.Finance.Util.Date.DayCount (calcPeriodDcf)
+import Daml.Finance.Util.Date.RollConvention (addPeriod)
+import Prelude hiding (key)
+
+-- | Create claims from swapStream that describes a fix coupon stream.
+calculateClaimsFromFixSwapStream : SwapStream -> PeriodicSchedule -> [SchedulePeriod] -> Bool -> Bool -> Deliverable -> Update TaggedClaim
+calculateClaimsFromFixSwapStream s periodicSchedule streamSchedule useAdjustedDatesForDcf issuerPaysLeg currency = do
+  let f = fromSome s.calculationPeriodAmount.calculation.fixedRateSchedule
+  pure $ createFixRatePaymentClaims streamSchedule periodicSchedule useAdjustedDatesForDcf f.initialValue issuerPaysLeg s.calculationPeriodAmount.calculation.dayCountFraction currency
+
+-- | Rule when one specific floating rate is provided for a stub period.
+getSingleStubRateFromFpml : DA.Internal.Record.HasField "floatingRateIndex" r o => r -> Optional (Observation t x o)
+getSingleStubRateFromFpml floatingRate = do
+  Some (Observe floatingRate.floatingRateIndex)
+
+-- | Linearly interpolates two rates within a period, as specified in https://www.isda.org/a/aWkgE/Linear-interpolation-04022022.pdf
+getInterpolatedStubRateFromFpml : (DA.Internal.Record.HasField "adjustedEndDate" r1 Date, DA.Internal.Record.HasField "adjustedStartDate" r1 Date, DA.Internal.Record.HasField "floatingRateIndex" r2 o, DA.Internal.Record.HasField "floatingRateIndex" r3 o, DA.Internal.Record.HasField "indexTenor" r2 (Optional Period), DA.Internal.Record.HasField "indexTenor" r3 (Optional Period)) => r2 -> r3 -> r1 -> HolidayCalendarData -> BusinessDayConventionEnum -> Optional (Observation t Decimal o)
```

Comment:
```suggestion
getInterpolatedStubRateFromFpml : => StubFloatingRate -> StubFloatingRate -> SchedulePeriod -> HolidayCalendarData -> BusinessDayConventionEnum -> Optional (Observation Time Decimal Observable)
```
seems to be a more readable signature.

Given that you seem to be using very often `Observation Time Decimal Observable`, you can define a type synonym at the beginning of the file, e.g. `type O = Observation Time Decimal Observable`